src/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs

```diff
@@ -148,76 +88,64 @@ internal void BeginReadLine()
-
-            try
-            {
-                byteLen = t.GetAwaiter().GetResult();
-            }
-            catch (IOException)
-            {
-                // We should ideally consume errors from operations getting cancelled
-                // so that we don't crash the unsuspecting parent with an unhandled exc. 
-                // This seems to come in 2 forms of exceptions (depending on platform and scenario), 
-                // namely OperationCanceledException and IOException (for errorcode that we don't 
-                // map explicitly).   
-                byteLen = 0; // Treat this as EOF
-            }
-            catch (OperationCanceledException)
-            {
-                // We should consume any OperationCanceledException from child read here  
-                // so that we don't crash the parent with an unhandled exc
-                byteLen = 0; // Treat this as EOF
-            }
-
-            if (byteLen == 0)
+            while (true)
             {
-                // We're at EOF, we won't call this function again from here on.
-                lock (_messageQueue)
+                try
                 {
-                    if (_sb.Length != 0)
-                    {
-                        _messageQueue.Enqueue(_sb.ToString());
-                        _sb.Length = 0;
-                    }
-                    _messageQueue.Enqueue(null);
-                }
+                    int bytesRead = await _stream.ReadAsync(_byteBuffer, 0, _byteBuffer.Length, _cts.Token).ConfigureAwait(false);
+                    if (bytesRead == 0)
+                        break;
 
-                try
+                    int charLen = _decoder.GetChars(_byteBuffer, 0, bytesRead, _charBuffer, 0);
+                    _sb.Append(_charBuffer, 0, charLen);
+                    MoveLinesFromStringBuilderToMessageQueue();
+                }
+                catch (IOException)
                 {
-                    // UserCallback could throw, we should still set the eofEvent 
-                    FlushMessageQueue();
+                    // We should ideally consume errors from operations getting cancelled
+                    // so that we don't crash the unsuspecting parent with an unhandled exc.
+                    // This seems to come in 2 forms of exceptions (depending on platform and scenario),
+                    // namely OperationCanceledException and IOException (for errorcode that we don't
+                    // map explicitly).
+                    break; // Treat this as EOF
                 }
-                finally
+                catch (OperationCanceledException)
                 {
-                    _eofEvent.Set();
+                    // We should consume any OperationCanceledException from child read here
+                    // so that we don't crash the parent with an unhandled exc
+                    break; // Treat this as EOF
                 }
+
+                FlushMessageQueue();
```

Comment:
The current .NET Core behavior is an oversight and not intentional. I don't think we should swallow the exception and should propagate it via a thread pool thread (desktop behavior). This pattern seems more intuitive. @stephentoub Thoughts?